homeassistant/components/subaru/__init__.py

```diff
@@ -0,0 +1,250 @@
+    """Set up Subaru from a config entry."""
+    hass.data.setdefault(DOMAIN, {})
+    config = entry.data
+    websession = aiohttp_client.async_get_clientsession(hass)
+    date = datetime.now().strftime("%Y-%m-%d")
+    device_name = "Home Assistant: Added " + date
+    try:
+        controller = SubaruAPI(
+            websession,
+            config[CONF_USERNAME],
+            config[CONF_PASSWORD],
+            config[CONF_DEVICE_ID],
+            config[CONF_PIN],
+            device_name,
+            update_interval=entry.options.get(
+                CONF_HARD_POLL_INTERVAL, DEFAULT_HARD_POLL_INTERVAL
+            ),
+            fetch_interval=entry.options.get(CONF_SCAN_INTERVAL, DEFAULT_SCAN_INTERVAL),
+        )
+        if not await controller.connect():
+            _LOGGER.error("Failed to connect")
+            return False
+    except SubaruException as ex:
+        _LOGGER.error("Unable to communicate with Subaru API: %s", ex.message)
+        return False
+
+    vehicle_info = {}
+    for vin in controller.get_vehicles():
+        vehicle_info[vin] = get_vehicle_info(controller, vin)
+
+    async def async_update_data():
+        """Fetch data from API endpoint."""
+        try:
+            async with async_timeout.timeout(30):
+                data = await subaru_update(vehicle_info, controller)
+            return data
+        except SubaruException as err:
+            raise UpdateFailed(f"Error communicating with API: {err}")
+
+    coordinator = DataUpdateCoordinator(
+        hass,
+        _LOGGER,
+        name="subaru_data",
+        update_method=async_update_data,
+        update_interval=timedelta(seconds=60),
+    )
+
+    await coordinator.async_refresh()
+
+    hass.data[DOMAIN][entry.entry_id] = {
+        "controller": controller,
+        "coordinator": coordinator,
+        "vehicles": vehicle_info,
+        "listener": [entry.add_update_listener(update_listener)],
+    }
+
+    for component in SUPPORTED_PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, component)
+                for component in SUPPORTED_PLATFORMS
+            ]
+        )
+    )
+    if unload_ok:
+        hass.data[DOMAIN].pop(entry.entry_id)
+    if len(hass.data[DOMAIN]) == 0:
+        hass.data.pop(DOMAIN)
+
+    return unload_ok
+
+
+async def update_listener(hass, config_entry):
+    """Update when config_entry options update."""
+    controller = hass.data[DOMAIN][config_entry.entry_id]["controller"]
+    controller.set_update_interval(config_entry.options.get(CONF_HARD_POLL_INTERVAL))
+    controller.set_fetch_interval(config_entry.options.get(CONF_SCAN_INTERVAL))
```

Comment:
You should use defaults in case these don't exist

Reply:
defaults added in 934d31e